Data/ByteString/Lazy/ReadInt.hs

```diff
@@ -0,0 +1,220 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE MultiWayIf #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+#include "MachDeps.h"
+
+-- This file is included by "Data.ByteString.ReadInt",
+-- "Data.ByteString.ReadInt64", and "Data.ByteString.Lazy.ReadInt64", after
+-- defining "BYTESTRING_STRICT" or "API_BITS_64" or both.  The modules are
+-- largely identical, except for the choice of ByteString type, and the
+-- the use of native words, or explicit 64-bits words for the accumulator.
+
+#ifndef API_BITS_64
+
+# ifdef BYTESTRING_STRICT
+module Data.ByteString.ReadInt
+# else
+module Data.ByteString.Lazy.ReadInt
+# endif
+    ( readInt
+    , readInt8
+    , readInt16
+    , readInt32
+    ) where
+
+#else
+
+# ifdef BYTESTRING_STRICT
+module Data.ByteString.ReadInt64
+# else
+module Data.ByteString.Lazy.ReadInt64
+# endif
+    ( readInt64 ) where
+
+#endif
+
+import qualified Data.ByteString.Internal as BI
+import qualified Data.ByteString.Unsafe as BU
+#ifdef BYTESTRING_STRICT
+import Data.ByteString as B
+#else
+import Data.Bits ((.|.))
+import Data.ByteString.Lazy as B
+import Data.ByteString.Lazy.Internal
+#endif
+import Data.Int
+import Data.Word
+import Foreign.Ptr (minusPtr, plusPtr)
+import Foreign.Storable (Storable(..))
+
+----- Public API
+
+#ifndef API_BITS_64
+type Accum = Word
+
+-- | Try to read a signed 'Int' value from the 'ByteString', returning
+-- @Just (val, str)@ on success, where @val@ is the value read and @str@ is the
+-- rest of the input string.  If the sequence of digits decodes to a value
+-- larger than can be represented by an 'Int', the returned value will be
+-- 'Nothing'.
+--
+-- 'readInt' does not ignore leading whitespace, the value must start
+-- immediately at the beginning of the input string.
+--
+-- ==== __Examples__
+-- >>> readInt "-1729 sum of cubes"
+-- Just (-1729," sum of cubes")
+-- >>> readInt "+1: readInt also accepts a leading '+'"
+-- Just (1, ": readInt also accepts a leading '+'")
+-- >>> readInt "not a decimal number")
+-- Nothing
+-- >>> readInt "12345678901234567890 overflows maxBound")
+-- Nothing
+-- >>> readInt "-12345678901234567890 underflows minBound")
+-- Nothing
+--
+readInt :: ByteString -> Maybe (Int, ByteString)
+{-# NOINLINE readInt #-}
+readInt = _readInt q
+  where q =
+# if WORD_SIZE_IN_BITS == 64
+            0xccccccccccccccc -- 0xccc_cccc_cccc_cccc
+# elif WORD_SIZE_IN_BITS == 32
```

Comment:
Please use `Data.Bits.finiteBitSize (0 :: Word)` instead of CPP. GHC is perfectly capable to evaluate `finiteBitSize` in compile time and prune dead branches.

Reply:
In that case the constant `0xccc_cccc_cccc_cccc` would then have to be computed I guess, since on 32-bit systems the overflowing literal, even if in a dead branch would generate a warning. If that's the case, I don't need `finiteBitSize`, just go back to `maxBound \`div\` 10`...

Reply:
Yep, ``maxBound `quot` 10`` is better from my perspective.